### PR TITLE
Fix detection of encryption for all users in a room

### PR DESCRIPTION
**Pull request by weeman1337**
fixes https://github.com/vector-im/element-web/issues/24995

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

element-web notes: When starting a DM, it is now ensured that the message can be encrypted for all users before encrypting the room.

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix detection of encryption for all users in a room ([\#10487](https://github.com/matrix-org/matrix-react-sdk/pull/10487)). Fixes vector-im/element-web#24995.

### CODE REVIEW

**Review by t3chguy — APPROVED** [review]

File: src/createRoom.ts [hunk 1]
```diff
@@ -398,16 +398,21 @@ export default async function createRoom(opts: IOpts): Promise<string | null> {
...
     try {
         const usersDeviceMap = await client.downloadKeys(userIds);
-        // { "@user:host": { "DEVICE": {...}, ... }, ... }
-        return Object.values(usersDeviceMap).every(
-            (userDevices) =>
-                // { "DEVICE": {...}, ... }
-                Object.keys(userDevices).length > 0,
-        );
+
+        // There are no devices at all.
```
Comment (richvdh) [inline src/createRoom.ts:402]: s/devices/users/.

I think this is a change of behaviour from before 3.69.0. Are we sure it's what's expected?

It's also inconsistent with the documentation on the function. If there are *no* users in the room (or rather, in the `userIds` list) then "there is at least one device that we can encrypt to" is trivially true.

File: src/createRoom.ts [hunk 2]
```diff
@@ -398,16 +398,21 @@ export default async function createRoom(opts: IOpts): Promise<string | null> {
...
+        if (usersDeviceMap.size === 0) return false;
+
+        for (const devices of usersDeviceMap.values()) {
+            if (devices.size === 0) {
+                return false;
```
Comment (richvdh) [inline src/createRoom.ts:407]: A comment would be useful here:

```suggestion
                // this user does not have any encryption-capable devices.
                return false;
```

File: src/createRoom.ts (hunk 1 above)
Comment (richvdh) [reply src/createRoom.ts:402]: > It's also inconsistent with the documentation on the function. If there are _no_ users in the room (or rather, in the `userIds` list) then "there is at least one device that we can encrypt to" is trivially true.

sorry, I misquoted.

If there are *no* users, then "for every user in a room, \<X>" is trivially true, for any \<X>.

Comment (weeman1337) [reply src/createRoom.ts:402]: Yes, this is a change. Not sure if it is correct. If I download the keys for a list of users and I do get an empty response, can I not encrypt to them? :thinking:

An empty `userIds` arg is a slightly separate case. If I do not specify a user, I cannot encrypt to them.

Comment (andybalaam) [reply src/createRoom.ts:402]: If a user has no devices, I don't think I want that to make me stop sending to a whole list of users. I can trivially do the right thing for that user, which is nothing.

Happy to be persuaded though.

Comment (weeman1337) [reply src/createRoom.ts:402]: @andybalaam do you mean a misc case here where some users have devices?

The discussion from above is about this case:
- Before: 0 users or 0 devices returned by `downloadKeys` → `true`
- Now: `false`

Comment (weeman1337) [reply src/createRoom.ts:402]: https://github.com/vector-im/element-web/issues/24998

Comment (richvdh) [reply src/createRoom.ts:402]: > If a user has no devices, I don't think I want that to make me stop sending to a whole list of users. I can trivially do the right thing for that user, which is nothing.

First of all, let's clarify. This code is relying on `downloadKeys` only returning devices which have encryption keys. So we're talking about *encryption-capable* devices here.

The whole point of this function is to check that all users in a given list have at least one encryption-capable device - which means we can safely encrypt a room which contains those users.

So, I disagree with you here @andybalaam: if any one of the users in the list has no (encryption-capable) devices, I do *not* want to start an encrypted room with the whole list.

Comment (andybalaam) [reply src/createRoom.ts:402]: What Rich said makes perfect sense to me.

**Pull request merged**